### PR TITLE
Nonlocal FFs, Lagrange polynomials and all commits related to the b->sll analysis

### PR DESCRIPTION
**Pull request by mreboud**

### CODE REVIEW

**Comment by mreboud** [comment]
I forgot that the names I used for the models based on Lagrange polynomials were so naive. I should probably change this for something like GRvDV2022.
I also think that GRvDV2021 can be removed because it is patently useless...

File: python/eos/plot/plotter.py [hunk 1]
```diff
@@ -671,9 +675,9 @@ def plot(self):
             ovalues_central = []
             ovalues_higher  = []
             for i in range(len(self.xvalues)):
-                lower   = np.percentile(self.samples[:, i], q=15.865)
-                central = np.percentile(self.samples[:, i], q=50.000)
-                higher  = np.percentile(self.samples[:, i], q=84.135)
+                lower, central, higher = self.plotter._weighted_quantiles(self.samples[:, i],
```
Comment (dvandyk) [inline python/eos/plot/plotter.py:678]: Nice, thank you!

File: eos/constraints/Bs-to-phi-charmonium.yaml [hunk 2]
```diff
@@ -56,6 +56,33 @@ B_s->phiJpsi::PolarizationAmplitude@LHCb:2014D:
...
+        - B_s->phipsi::delta_perp_long
+    kinematics:
+        [ {}, {}, {}, {} ]
+    options:
+        [ {psi: "psi(2S)"}, {psi: "psi(2S)"}, {psi: "psi(2S)"}, {psi: "psi(2S)"} ]
+    means: [ 0.264, 0.422, 3.67, 3.29 ]
+    sigma-stat-hi: [ 0.024, 0.014, 0.13, 0.43 ]
+    sigma-stat-lo: [ 0.023, 0.014, 0.18, 0.39 ]
+    sigma-sys: [ 0.002, 0.003, 0.03, 0.04 ]
+    correlations:
```
Comment (dvandyk) [inline eos/constraints/Bs-to-phi-charmonium.yaml:78]: Out of curiosity: is the LHCb result expressed in terms of standard deviations and correlation, or in terms of covariance?

File: eos/parameters/hadronic-matrix-elements.yaml [hunk 3]
```diff
@@ -4344,6 +4344,32 @@ groups:
               - [ "Re", "Im" ]
               - [ "0", "1", "2", "3", "4", "5" ]
               - [ "perp", "para", "long"]
+      'B->K^*ccbar::%1%_at_%2%_%3%@Lagrange' :
```
Comment (dvandyk) [inline eos/parameters/hadronic-matrix-elements.yaml:4359]: The qualified name does not contain the actual name of the quantity, does it? E.g., ``Re{H_perp(t0)}@Lagrange``?
Can a qualified name contain ``-1.0``?

File: eos/rare-b-decays/b-to-k-charmonium.cc [hunk 4]
```diff
@@ -76,7 +77,7 @@ namespace eos
             m_B(p["mass::B_" + opt_q.value()], u),
             tau_B(p["life_time::B_" + opt_q.value()], u),
             m_K(p["mass::K_" + opt_q.value()], u),
-            opt_nonlocal_formfactor(o, "nonlocal-formfactor", { "GvDV2020", "GRvDV2021" }, "GvDV2020"),
+            opt_nonlocal_formfactor(o, "nonlocal-formfactor", { "GvDV2020", "GRvDV2021", "Lagrange", "LagrangeUncertainty" }, "GvDV2020"),
```
Comment (dvandyk) [inline eos/rare-b-decays/b-to-k-charmonium.cc:80]: As discussed, ``Lagrange`` -> ``GRvDV2022``.
Remove ``GRvDV2021``?

File: eos/rare-b-decays/b-to-k-ll-gvdv2020.cc [hunk 5]
```diff
@@ -41,7 +41,7 @@ namespace eos
         f_K(p["decay-constant::K_" + o.get("q", "d")], *this),
         lambda_B_p_inv(p["B::1/lambda_B_p"], *this),
         q(o, "q", { "d", "u" }, "d"),
-        opt_nonlocal_formfactor(o, "nonlocal-formfactor", { "GvDV2020", "GRvDV2021", "naive" }, "GvDV2020"),
+        opt_nonlocal_formfactor(o, "nonlocal-formfactor", { "GvDV2020", "GRvDV2021", "naive", "Lagrange", "LagrangeUncertainty" }, "GvDV2020"),
```
Comment (dvandyk) [inline eos/rare-b-decays/b-to-k-ll-gvdv2020.cc:44]: Ditto.

File: eos/rare-b-decays/nonlocal-formfactors-p-to-p.cc [hunk 6]
```diff
@@ -752,6 +752,613 @@ namespace eos
...
+         * but using an ad hoc Lagrange polynomial.
+         */
+        template <typename Process_>
+        class Lagrange :
+            public NonlocalFormFactor<nff::PToP>
+        {
+            public:
+                std::shared_ptr<FormFactors<PToP>> form_factors;
+
+                //Polynomial expansion parameters
```
Comment (dvandyk) [inline eos/rare-b-decays/nonlocal-formfactors-p-to-p.cc:869]: Space.

File: eos/rare-b-decays/nonlocal-formfactors-p-to-p.cc [hunk 7]
```diff
@@ -752,6 +752,613 @@ namespace eos
...
+                    const complex<double> phi1 = -pow(2 * pow((4 * m_D02 - Q2) * (4 * m_D02 - s_0), 0.5) + 8 * m_D02 - Q2 - s_0, 0.5) /
+                                                (2 * pow((4 * m_D02 - Q2) * (4 * m_D02 - s_0), 0.5) + 8 * m_D02 + Q2 * (z - 1.) - s_0 * (z + 1.)); //(C7)
+                    const complex<double> phi2 = pow(m_B4 * power_of<4>(z - 1.) - 2 * m_B2 * power_of<2>(z - 1.) * (-16 * m_D02 * z + m_P2 * power_of<2>(z - 1.) + s_0 * pow(z + 1., 2)) +
+                                                power_of<2>(16 * m_D02 * z + m_P2 * power_of<2>(z - 1.) - s_0 * power_of<2>(z + 1.)), 0.5);//(C8)
+                    const complex<double> phi3 = pow(8 * m_D02 + 4 * pow(4 * m_D04 - s_0 * m_D02, 0.5) - s_0, 0.5) /
+                                                (-8 * m_D02 - 4 * pow(4 * m_D04 - s_0 * m_D02, 0.5) + s_0*(z+1.)); //(C9)
+                    const complex<double> phi4 = pow(s_0 * power_of<2>(z + 1.) - 16. * z * m_D02, -0.5); //(C10)
+
+                    return Nlambda * pow(1.+z, 0.5) * pow(1.-z, a-b+c+d-1.5) * pow(phi1, a) * pow(phi2, 0.5*b) * pow(phi3, c) * pow(phi4, d); //(C5)
```
Comment (dvandyk) [inline eos/rare-b-decays/nonlocal-formfactors-p-to-p.cc:985]: Please use spaces in arithmetic expressions.

File: eos/rare-b-decays/observables.cc [hunk 8]
```diff
@@ -531,6 +531,14 @@ namespace eos
                         Unit::None(),
                         &NonlocalFormFactorObservable<nff::BToKstar, nff::PToV>::abs_H_perp,
                         std::make_tuple("q2")),
+                make_observable("B->K^*::re_Hhat_perp(q2)",  R"(\mathrm{Re}\hat{\mathcal{H}}_\perp^{B \to K^*}(q^2))",
+                        Unit::None(),
```
Comment (dvandyk) [inline eos/rare-b-decays/observables.cc:543]: Double check that the Hhat functions are unit less. I doubt it.

File: eos/b-decays/b-to-kstar-nu-nu_TEST.cc [hunk 9]
```diff
@@ -81,10 +81,10 @@ class BToKDineutrinoTest :
                 const double eps = 1e-4;
 
                 TEST_CHECK_RELATIVE_ERROR(2.84988e-06, d.integrated_branching_ratio( 0.00,  8.00), eps);
-                TEST_CHECK_RELATIVE_ERROR(4.07784e-06, d.integrated_branching_ratio( 8.00, 16.00), eps);
```
Comment (dvandyk) [inline eos/b-decays/b-to-kstar-nu-nu_TEST.cc:84]: Thanks!

File: eos/rare-b-decays/b-to-kstar-ll-gvdv2020_TEST.cc [hunk 10]
```diff
@@ -255,3 +255,199 @@ class BToKstarDileptonJavierTest :
...
+
+
+            // Javier's results
+            // {BR,FL,AFB,P1,P2,P3,P4p,P5p,P6p,P8p,S3,S4,S5,S7,S8,S9} =
+            // {0.669443, 0.506906, 0.167357, -0.0428339, -0.259782, 0.000138555, 0.920583, -0.831674, -0.00743048, 0.0077058, -0.00919817, 0.214768, -0.388051, 0.00346699, -0.00179773, -0.0000595069}  
+
+            TEST_CHECK_RELATIVE_ERROR(
+                 Observable::make("B->K^*ll::dBR/ds", p, k, oo)->evaluate(),    0.669443e-07, eps );
+            TEST_CHECK_RELATIVE_ERROR(
+                 Observable::make("B->K^*ll::F_L(q2)", p, k, oo)->evaluate(),   0.542507,     eps );
```
Comment (dvandyk) [inline eos/rare-b-decays/b-to-kstar-ll-gvdv2020_TEST.cc:422]: The discrepancy to Javier's results in F_L requires a comment. If typo in Javier's results (as discussed), please update the comment above.

File: eos/rare-b-decays/b-to-kstar-ll-gvdv2020_TEST.cc [hunk 11]
```diff
@@ -255,3 +255,199 @@ class BToKstarDileptonJavierTest :
...
+            TEST_CHECK_RELATIVE_ERROR(c.differential_j_6s(q2), -3.0/4.0*6.47412e-21,  eps);
+            TEST_CHECK_NEARLY_EQUAL(c.differential_j_6c(q2),  0.0,  1e-22);
+            TEST_CHECK_RELATIVE_ERROR(c.differential_j_7(q2),   3.0/4.0*1.00589e-22,  eps);
+            TEST_CHECK_RELATIVE_ERROR(c.differential_j_8(q2),  -3.0/4.0*5.21581e-23,  eps);
+            TEST_CHECK_RELATIVE_ERROR(c.differential_j_9(q2),  -3.0/4.0*1.72649e-24,  eps);
+
+
+            // Javier's results
+            // {BR,FL,AFB,P1,P2,P3,P4p,P5p,P6p,P8p,S3,S4,S5,S7,S8,S9} =
+            // {0.669443, 0.506906, 0.167357, -0.0428339, -0.259782, 0.000138555, 0.920583, -0.831674, -0.00743048, 0.0077058, -0.00919817, 0.214768, -0.388051, 0.00346699, -0.00179773, -0.0000595069}
```
Comment (dvandyk) [inline eos/rare-b-decays/b-to-kstar-ll-gvdv2020_TEST.cc:417]: Please use a linebreak or two.

File: eos/rare-b-decays/observables.cc [hunk 12]
```diff
@@ -1337,76 +1348,160 @@ namespace eos
...
                         )"),
 
+                make_expression_observable("B->K^*ll::S_2c(q2)", R"(S_{2c}^{\bar{B}\to \bar{K}^*\ell^+\ell^-}(q^2))",
+                        Unit::None(),
+                        R"(
+                        2.0 / 3.0 * (<<B->K^*ll::J_2c(q2);cp-conjugate=false>> + <<B->K^*ll::J_2c(q2);cp-conjugate=true>>)
+                                  / <<B->K^*ll::Gamma(q2)>>
+                        )"),
+
+                make_expression_observable("B->K^*ll::S_3(q2)", R"(S_3^{\bar{B}\to \bar{K}^*\ell^+\ell^-}(q^2))",
```
Comment (dvandyk) [inline eos/rare-b-decays/observables.cc:1456]: Why is ``S_3(q2)`` barging the queue here? The others are **after** their respective integrated quantities.

File: eos/rare-b-decays/observables.cc [hunk 13]
```diff
@@ -1337,76 +1348,160 @@ namespace eos
...
+                        )"),
+
                 make_expression_observable("B->K^*ll::S_3", R"(S_3(\bar{B}\to \bar{K}^*\ell^+\ell^-))",
                         Unit::None(),
                         R"(
                         2.0 / 3.0 * (<<B->K^*ll::J_3;cp-conjugate=false>> + <<B->K^*ll::J_3;cp-conjugate=true>>)
                                   / <<B->K^*ll::Gamma>>
                         )"),
 
+                make_expression_observable("B->K^*ll::S_4(q2)", R"(S_4^{\bar{B}\to \bar{K}^*\ell^+\ell^-}(q^2))",
```
Comment (dvandyk) [inline eos/rare-b-decays/observables.cc:1470]: Ditto.

File: eos/rare-b-decays/observables.cc [hunk 14]
```diff
@@ -1337,76 +1348,160 @@ namespace eos
...
+                        )"),
+
                 make_expression_observable("B->K^*ll::S_4", R"(S_4(\bar{B}\to \bar{K}^*\ell^+\ell^-))",
                         Unit::None(),
                         R"(
                         2.0 / 3.0 * (<<B->K^*ll::J_4;cp-conjugate=false>> + <<B->K^*ll::J_4;cp-conjugate=true>>)
                                   / <<B->K^*ll::Gamma>>
                         )"),
 
+                make_expression_observable("B->K^*ll::S_5(q2)", R"(S_5^{\bar{B}\to \bar{K}^*\ell^+\ell^-}(q^2))",
```
Comment (dvandyk) [inline eos/rare-b-decays/observables.cc:1484]: Ditto.

File: eos/constraints/Bs-to-phi-charmonium.yaml (hunk 2 above)
Comment (mreboud) [reply eos/constraints/Bs-to-phi-charmonium.yaml:78]: image

File: eos/parameters/hadronic-matrix-elements.yaml (hunk 3 above)
Comment (mreboud) [reply eos/parameters/hadronic-matrix-elements.yaml:4359]: Hmm good point, will change that.
No, hence the `m1`...

File: eos/rare-b-decays/b-to-kstar-ll-gvdv2020_TEST.cc (hunk 10 above)
Comment (mreboud) [reply eos/rare-b-decays/b-to-kstar-ll-gvdv2020_TEST.cc:422]: There was indeed a typo in Javier's code, I found the discussion, I corrected the value but not the comment!

File: eos/rare-b-decays/observables.cc (hunk 8 above)
Comment (mreboud) [reply eos/rare-b-decays/observables.cc:543]: They are: Blaschke factors, outer functions and H are dimensionless

**Pull request merged**